Initial version of EMF Support for Event-B extension containment

tshoang opened the pull request:
This is the first attempt for Event-B extension containment.

tshoang committed:
Ignore files

tshoang committed:
Initial version of extension containments

tshoang committed:
Merge branch 'master' of https://github.com/eventB-Soton/EMF_Containment

tshoang committed:
Make the ac.soton.eventb.emf.containment an XText project (required for used in XEvent-B)

cfsnook commented on ac.soton.eventb.emf.containment.edit/plugin.xml:
```diff
@@ -0,0 +1,35 @@
+
+   <extension point="org.eclipse.emf.edit.childCreationExtenders">
+      <extender
+            uri="http://emf.eventb.org/models/core/2014"
```
change to http://emf.eventb.org/models/core/machine/2014
otherwise edit providers will not offer child 'Contains Extension's correctly
(unfortunately this is generated incorrectly due to the core/machine structure of the Event B EMF metamodel)

cfsnook commented on ac.soton.eventb.emf.containment.edit/plugin.xml:
```diff
@@ -0,0 +1,35 @@
+            class="ac.soton.eventb.emf.containment.provider.ContainmentItemProviderAdapterFactory$CoreChildCreationExtender"/>
+   </extension>
+
+</plugin>
```
We could also provide an extension for extension point org.eventb.emf.persistence.synchroniser
This would allow 'Contains Extension's to be persisted in Rodin Machines in case we need it
(see Machine Inclusion for example)

tshoang replied:
Do we actually store Machine Inclusion in the Rodin Machines? I will take a look at this.

cfsnook replied:
Yes, we use Machine Inclusion without XEventB (e.g. CODA) so we need to persist Machine Inclusion into Rodin Machines. Maybe we will need to do the same with Contains.. however,  adding this extension introduces a dependency on org.eventb.emf.persistence. Maybe it is better to add the extension elsewhere? E.g. in a new plugin ac.soton.eventb.emf.containment.persistence.

cfsnook commented on ac.soton.eventb.emf.containment.editor/.classpath:
```diff
@@ -0,0 +1,7 @@
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="output" path="bin"/>
+</classpath>
```
Do we need this editor? It is an editor (and wizard) for a resource with a 'Contains Extension' in it.  However, the intention of 'Contains Extension'  is to add a new clause in a Machine.. not to create a new model root. 
(Note that the Rose editor will handle this extension using the new edit providers).

tshoang replied:
We do not really. I just lazy and say "Generate All". I will remove it.

cfsnook commented on ac.soton.eventb.emf.containment/model/containment.ecore:
```diff
@@ -0,0 +1,11 @@
+    <eStructuralFeatures xsi:type="ecore:EAttribute" name="names" upperBound="-1"
+        eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString" defaultValueLiteral=""/>
+  </eClassifiers>
+</ecore:EPackage>
```
(For consideration only - could leave as is) 
The meta-model names feel a bit cumbersome & technical..
ContainsExtension -> Containment, or ContainedModel?
extension -> model?
names -> instances or instanceNames?

tshoang replied:
I actually started with ContainedModel and changed to ContainsExtension. It is the extension that we want to add to the machine so I think extension is appropriate (and we include AbstractExtension anyway). The reason why it is not ContainedExtension is too be consistent with RefinesMachine and SeesContext. But maybe ContainedExtension is better.
The reason for using names is because I want it to be "element name" so that there cannot be name clash. However, It probably better to have a single name, instead of "names". In fact, I think I should try to write the XText grammar so that we can do this.

cfsnook replied:
but the names attribute is multiplicity many - doesn't this mean we can include multiple (named) instances of the extension?

tshoang replied:
This is because in the XText grammar, we have
  contains
    a,b : ExtensionA
    c,d: ExtensionB
 which at the moment will create
   ContainsExtension{extension = ExtensionA, names=[a,b]}
   ContainsExtension{extension = ExtensionB, names=[c,d]}
 What is desirable (I'm working on it)
 is to have 
   ContainsExtension{name = a, extension = ExtensionA}
   ContainsExtension{name = b, extension = ExtensionA}
   ContainsExtension{name = c, extension = ExtensionB}
   ContainsExtension{name = d, extension = ExtensionB}

cfsnook replied:
Ok, yes i prefer the desirable option too. It is more flexible in case we need more attributes for the instantiation of an extension in the future.

tshoang replied:
Unfortunately, I do not know how to write the XText grammar to do what we want. Maybe @dd4g12 can shed some light on this, but as far as I can see, it is not (yet) possible. So the options left for us are:
1. Keep the current EMF Containment and the current grammar. Due to the mandatory of `extensionID` attribute of `AbstractExtension`, we have to have the `contains` clause as follows
```
 contains
    a b : ExtensionA
 contains
    c d : ExtensionB
```
2. Change the EMF Containment to single `name` attribute and adapt the grammar accordingly. However, due the the `extensionID`, we will have something like this
```
 contains
    a : ExtensionA
 contains
    b : ExtensionA
 contains
    c: ExtensionB
 contains
    d: ExtensionB
```

cfsnook replied:
Note that extensionID is no longer mandatory (in the new prototype release of Event-B EMF)

cfsnook reviewed (changes requested):
The extension point "org.eclipse.emf.edit.childCreationExtenders" needs to be fixed in the edit plugin.xml.
The other comments are suggestions

dd4g12 commented:
Yes it is machine in the inclusion plugin.

 

Dana

 

From: Thái Sơn Hoàng <notifications@github.com>
Reply-To: eventB-Soton/EMF_Containment <reply@reply.github.com>
Date: Wednesday, 27 September 2017 at 14:52
To: eventB-Soton/EMF_Containment <EMF_Containment@noreply.github.com>
Cc: dd4g12 <dd4g12@ecs.soton.ac.uk>, Review requested <review_requested@noreply.github.com>
Subject: Re: [eventB-Soton/EMF_Containment] Initial version of EMF Support for Event-B extension containment (#2)

 

@tshoang commented on this pull request.

In ac.soton.eventb.emf.containment.edit/plugin.xml:
> +            class="ac.soton.eventb.emf.containment.provider.ContainmentItemProviderAdapterFactory"
+            supportedTypes=
+              "org.eclipse.emf.edit.provider.IEditingDomainItemProvider
+               org.eclipse.emf.edit.provider.IStructuredItemContentProvider
+               org.eclipse.emf.edit.provider.ITreeItemContentProvider
+               org.eclipse.emf.edit.provider.IItemLabelProvider
+               org.eclipse.emf.edit.provider.IItemPropertySource"/>
+   </extension>
+
+   <extension point="org.eclipse.emf.edit.childCreationExtenders">
+      <extender
+            uri="http://emf.eventb.org/models/core/2014"
+            class="ac.soton.eventb.emf.containment.provider.ContainmentItemProviderAdapterFactory$CoreChildCreationExtender"/>
+   </extension>
+
+</plugin>
Do we actually store Machine Inclusion in the Rodin Machines? I will take a look at this.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub, or mute the thread.

{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/eventB-Soton/EMF_Containment","title":"eventB-Soton/EMF_Containment","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/eventB-Soton/EMF_Containment"}},"updates":{"snippets":[{"icon":"PERSON","message":"@tshoang commented on #2"}],"action":{"name":"View Pull Request","url":"https://github.com/eventB-Soton/EMF_Containment/pull/2#discussion_r141328159"}}}

tshoang committed:
Fixed the extension point "org.eclipse.emf.edit.childCreationExtenders" in the *edit* plugin.

Colin: change to http://emf.eventb.org/models/core/machine/2014
otherwise edit providers will not offer child 'Contains Extension's correctly
(unfortunately this is generated incorrectly due to the core/machine structure of the Event B EMF metamodel)

tshoang committed:
Removed unnecessary editor plugin

tshoang commented:
@cfsnook Can you please check this.
